### PR TITLE
Better diffs

### PR DESCRIPTION
**Pull request by mwsteb**
This makes diffing easy as pie. All we gotta do is implement the `#to_h` method in all the integrations’ base models (teams/groups).

Things still to be done:
- use new diffs in sync and log
- clean up old rubbish
- change how we handle completely new team (one that’s in permissions repo, but not yet on github): during syncing. check if the team exists first and if not - create it (empty). Then just add all the info as if it were there from the beginning. See? Easy.

### CODE REVIEW

File: app/models/github_integration/team.rb [hunk 1]
```diff
@@ -18,5 +36,15 @@ def to_yaml
         repos: repos || [],
       }.stringify_keys.to_yaml
     end
+
+    private
```
Comment (netguru-hound) [inline app/models/github_integration/team.rb]: Useless `private` access modifier.

File: spec/models/base_diff_spec.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
+require 'rails_helper'
+
+RSpec.describe BaseDiff do
+  let(:team_alfa) { GithubIntegration::Team.new('Alfa', %w(old-member), %w(old-repo), 'push') }
+  let(:team_beta) { GithubIntegration::Team.new('Beta', %w(older-member), %w(older-repo), 'push') }
+  let(:api_team_alfa) { GithubIntegration::Team.new('Alfa', %w(new-member), %w(new-repo), 'pull') }
+  let(:api_team_delta) { GithubIntegration::Team.new('Delta', %w(new-member), %w(new-repo), 'push') }
```
Comment (netguru-hound) [inline spec/models/base_diff_spec.rb]: Line is too long. [101/100]

File: app/models/github_integration/actions/get_teams_with_assoc.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    class GetTeamsWithAssoc
+      pattr_initialize :client
+
+      def now!
+        list_teams
+      end
+
+      private
+
+      def list_teams
```
Comment (netguru-hound) [inline app/models/github_integration/actions/get_teams_with_assoc.rb]: Assignment Branch Condition size for list_teams is too high. [17/15]

File: app/models/github_integration/actions/sync.rb [hunk 4]
```diff
@@ -1,81 +1,69 @@
...
+      pattr_initialize(:gh_api, :diff, :names_and_ids) do
+        @remove = diff.remove_hash
+        @add = diff.add_hash
       end
+      attr_reader :remove, :add
 
-      def now!(diff)
-        sync(diff)
+      def now!
+        add.each { |team_name, changes| sync_add_team(team_name.downcase, team_changes) }
```
Comment (netguru-hound) [inline app/models/github_integration/actions/sync.rb]: Unused block argument - `changes`. If it's necessary, use `_` or `_changes` as an argument name to indicate that it won't be used.

File: app/models/github_integration/actions/sync.rb [hunk 5]
```diff
@@ -1,81 +1,69 @@
...
+        remove.each { |team_name, changes| sync_remove_team(team_name.downcase, team_changes) }
```
Comment (netguru-hound) [inline app/models/github_integration/actions/sync.rb]: Unused block argument - `changes`. If it's necessary, use `_` or `_changes` as an argument name to indicate that it won't be used.

File: app/models/github_integration/sync_job.rb [hunk 6]
```diff
@@ -1,7 +1,7 @@
 module GithubIntegration
   class SyncJob
-    def perform(api, diff)
-      ping_messages { GithubIntegration::Actions::Sync.new(api).now!(diff) }
+    def perform(api, diff, gh_teams)
```
Comment (netguru-hound) [inline app/models/github_integration/sync_job.rb]: Unused method argument - `gh_teams`. If it's necessary, use `_` or `_gh_teams` as an argument name to indicate that it won't be used.

**Comment by lubieniebieski** [comment]
I'm closing this for now - it's a good direction but we weren't able to complete it on time:(

**Pull request closed without merging**